storage: avoid lease and range transfers to a draining node

asubiotto opened the pull request:
Extends StoreDescriptor with a draining field, so that no store that is known
to be draining is the target of a lease or range transfer. This information is
forcefully gossipped when draining but is a best effort attempt to avoid
transfers. Future work will add a rejection of lease and range transfers on a
draining node.

Part of `drain-leases` work proposed in #10765.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/11724)
<!-- Reviewable:end -->

petermattis commented:
Mostly looks good. 

---

Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator_test.go, line 311 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXqgFsHbACrbn8zhIDr:-KXqgFsI2WS3k9-urszO:b-tddxc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/allocator_test.go#L311)):*
> ```Go
> 
> 		sameDCStores[0].Draining = true
> 		defer func() { sameDCStores[0].Draining = false }()
> ```

Tweaking a global is a bit ugly.  Perhaps create a local slice of store descriptors here:

```
	var stores []*roachpb.StoreDescriptor
	for i := 1; i <= 3; i++ {
		stores = append(stores, &roachpb.StoreDescriptor{
			StoreID:  roachpb.StoreID(i),
			Node:     roachpb.NodeDescriptor{NodeID: roachpb.NodeID(i)},
			Capacity: roachpb.StoreCapacity{Capacity: 100, Available: 200},
			Draining: i == 1,
		})
	}
```

---

*[pkg/storage/store.go, line 911 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXqgw_JC1qWNE5iUEnC:-KXqgw_JC1qWNE5iUEnD:b-4iqs0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/store.go#L911)):*
> ```Go
> 	// Force a gossip of the draining mode.
> 	select {
> 	case <-s.cfg.Gossip.Connected:
> ```

Why is it necessary to check this? Some test?

---

*[pkg/storage/store.go, line 912 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXqgmkgp5hZLaMqFEMR:-KXqgmkgp5hZLaMqFEMS:be39ve9) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/store.go#L912)):*
> ```Go
> 	select {
> 	case <-s.cfg.Gossip.Connected:
> 		_ = s.GossipStore(s.AnnotateCtx(context.Background()))
> ```

Don't ignore the error here: log it.

---

*[pkg/storage/store_pool.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXqh7QB0ogcl-H6V-W7:-KXqh7QB0ogcl-H6V-W8:beojlye) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/store_pool.go#L125)):*
> ```Go
> 
> 	if sd.desc.Draining {
> 		return storeStatusDead
> ```

We rebalance away from dead stores which I don't think is desired here. Better to return `storeStatusThrottled`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-:-KXqhJH5zs7tR2rIfun7:b-z8proa)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.proto, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXybJ9JDny-2xiunVGr:-KXybJ9KfsVECfnwherO:bvwvzd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/roachpb/metadata.proto#L105)):*
> ```protobuf
>   optional NodeDescriptor node = 3 [(gogoproto.nullable) = false];
>   optional StoreCapacity capacity = 4 [(gogoproto.nullable) = false];
>   optional bool draining = 5 [(gogoproto.nullable) = false];
> ```

Add comments about how this proto is used both on-disk and in gossip, and the capacity and draining fields are not actually written to disk. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator_test.go, line 300 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXzAL5FKJ1X8FIydEx9:-KXzAL5FKJ1X8FIydExA:b2g04o) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/allocator_test.go#L300)):*
> ```Go
> 
> // TestAllocatorDrainingStore ensures that the allocator never attempts to
> // allocate a new replica on top of a draining store.
> ```

s/top of/to/

---

*[pkg/storage/allocator_test.go, line 323 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXzAWs_zoRGqOywo1Um:-KXzAWs_zoRGqOywo1Un:bpoa58o) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/allocator_test.go#L323)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		if result.Node.NodeID != 2 || result.StoreID != 2 {
> ```

this is awfully opaque and also overspecified for what this test is doing. maybe just check that `result.Draining` is false?

---

*[pkg/storage/store.go, line 912 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXz9jyhV3IpMISGnTqY:-KXz9jyhV3IpMISGnTqZ:b-r3jrg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/store.go#L912)):*
> ```Go
> 	select {
> 	case <-s.cfg.Gossip.Connected:
> 		_ = s.GossipStore(s.AnnotateCtx(context.Background()))
> ```

why Background() and not TODO()?

---

*[pkg/storage/store_pool_test.go, line 368 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KXzA9qMogjpKFmcy1K6:-KXzA9qMogjpKFmcy1K7:b-sq87ht) ([raw file](https://github.com/cockroachdb/cockroach/blob/58ad831dc07106cf24b1da9b765a084d5f155ed5/pkg/storage/store_pool_test.go#L368)):*
> ```Go
> 
> 	// Set drainingStore as draining.
> 	drainingStore.Draining = true
> ```

why can't this be in the `StoreDescriptor` literal?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

Commit:
storage: avoid lease and range transfers to a draining node

Extends Liveness proto with a draining field. When a node starts
draining, it will attempt to update its liveness record with an updated
draining field. This is then taken into account by StorePool to
determine valid stores to transfer leases/ranges to.

asubiotto commented:
Rewrote the PR to use node liveness instead of the store descriptor proto. PTAL.

---

Review status: 0 of 6 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-:-KcOnIZDGviXB2FQ7b7X:bw46da6)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 18 of 18 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/storage/liveness.proto, line 34 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOnngvdSW-9dWR9eRH:-KcOnngvdSW-9dWR9eRI:b-9guhfr) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/liveness.proto#L34)):*
> ```protobuf
>   // The timestamp at which this liveness record expires.
>   util.hlc.Timestamp expiration = 3 [(gogoproto.nullable) = false];
>   bool draining = 4;
> ```

nullable = false?

---

*[pkg/storage/node_liveness.go, line 174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOqHfhbqkLD6blWnyU:-KcOqHfia-xm73CONO4r:bm3cstg) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L174)):*
> ```Go
> 			log.Errorf(ctx, "unexpected error getting liveness: %s", err)
> 		}
> 		var newLiveness Liveness
> ```

how about:
```
newLiveness := Liveness{
	NodeID:   nl.gossip.NodeID.Get(),
	Epoch:    1,
}
if liveness != nil {
  newLiveness = *liveness
}
newLiveness.Draining = drain
```

perhaps also do this in the other location that follows a similar pattern

---

*[pkg/storage/node_liveness.go, line 185 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOqgAntdFxnzYYRA7x:-KcOqgAntdFxnzYYRA7y:b-qbnmmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L185)):*
> ```Go
> 			newLiveness.Draining = drain
> 		}
> 		update := func(l Liveness) {
> ```

perhaps you can make this a proper method and reuse it in the other places that duplicate this pattern

---

*[pkg/storage/node_liveness.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOpv1PUNsaOG7EdtSi:-KcOpv1PUNsaOG7EdtSj:bqwspn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L192)):*
> ```Go
> 		if err := nl.updateLiveness(ctx, &newLiveness, liveness, func(actual Liveness) error {
> 			update(actual)
> 			return errors.New("failed to update liveness record")
> ```

this doesn't seem right; if the liveness record already present already has the right draining value, why return an error?

this isn't tested at the unit level, either.

---

*[pkg/storage/node_liveness_test.go, line 462 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoJ3hfaKUNb0shQHq:-KcOoJ3iqWonHz99JbtK:by17d0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness_test.go#L462)):*
> and the node will not be a target for any transfers once
> // other nodes are aware of its draining state.

This is not directly tested.

---

*[pkg/storage/node_liveness_test.go, line 478 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoOiBxpWLbAZ0wTJ_:-KcOoOiBxpWLbAZ0wTJa:b9yazdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness_test.go#L478)):*
> ```Go
> 	testutils.SucceedsSoon(t, func() error {
> 		for i, sp := range mtc.storePools {
> 			if _, alive, _ := sp.GetStoreList(roachpb.RangeID(0)); alive != expectedLive {
> ```

this can just be `sp.GetStoreList(0)`, I think.

---

*[pkg/storage/store_pool.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoAB4IDs33satAK6U:-KcOoAB4IDs33satAK6V:bo99h9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/store_pool.go#L69)):*
> ```Go
> 			return false
> 		}
> 		deadAsOf := liveness.Expiration.GoTime().Add(threshold)
> ```

nit: you can avoid this if `liveness.Draining`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
It doesn't look like a restarted server will ever take itself out of draining mode. 

---

Reviewed 17 of 18 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/storage/liveness.proto, line 34 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOnngvdSW-9dWR9eRH:-KcTokEsXZWjupMQ7lFo:bf8ok2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/liveness.proto#L34)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nullable = false?
</blockquote></details>

This file is proto3, so primitives are always non-nullable.

---

*[pkg/storage/node_liveness.go, line 158 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcTppmC0GQD_95l8HXk:-KcTppmC0GQD_95l8HXl:b-mvl1tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L158)):*
> ```Go
> 	// in which heartbeats are resumed after SetDraining(true) has paused them
> 	// by a previous call to SetDraining(false).
> 	nl.drainSem <- struct{}{}
> ```

This looks equivalent to a plain `Mutex`. The others are channels because we use `select` and a `Context` for cancellation. Either change this to a mutex or make this method take a `Context` as well (probably the latter since you use `context.Background` below)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-:-KcTrsLeHnwUxTRgQFc_:b-puvatu)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
You're right. Fixed and tested.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 158 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcTppmC0GQD_95l8HXk:-KcUaHIqUgIim_PG1gc-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L158)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This looks equivalent to a plain `Mutex`. The others are channels because we use `select` and a `Context` for cancellation. Either change this to a mutex or make this method take a `Context` as well (probably the latter since you use `context.Background` below)
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOqHfhbqkLD6blWnyU:-KcTj-iDrI3CUfc37wpG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L174)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how about:
```
newLiveness := Liveness{
	NodeID:   nl.gossip.NodeID.Get(),
	Epoch:    1,
}
if liveness != nil {
  newLiveness = *liveness
}
newLiveness.Draining = drain
```

perhaps also do this in the other location that follows a similar pattern
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 185 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOqgAntdFxnzYYRA7x:-KcTj2EpvQ8R_e5Hh3_J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L185)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

perhaps you can make this a proper method and reuse it in the other places that duplicate this pattern
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOpv1PUNsaOG7EdtSi:-KcUcR_ewy9YtjamH_O6:bvs1ogk) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this doesn't seem right; if the liveness record already present already has the right draining value, why return an error?

this isn't tested at the unit level, either.
</blockquote></details>

Changed it to break out of the loop if the draining value is right. What exactly is not tested at the unit level?

---

*[pkg/storage/node_liveness_test.go, line 462 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoJ3hfaKUNb0shQHq:-KcUaTI89Psq6xMI7Nvx:b-j8kn91) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness_test.go#L462)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> and the node will not be a target for any transfers once
> // other nodes are aware of its draining state.

This is not directly tested.
</blockquote></details>

Modified to check presence/absence of the draining node's ID.

---

*[pkg/storage/node_liveness_test.go, line 478 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoOiBxpWLbAZ0wTJ_:-KcUair0ApK0yF7vXdJJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness_test.go#L478)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this can just be `sp.GetStoreList(0)`, I think.
</blockquote></details>

Done.

---

*[pkg/storage/store_pool.go, line 69 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoAB4IDs33satAK6U:-KcUbC7QQ04qlnQgZyfz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/store_pool.go#L69)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: you can avoid this if `liveness.Draining`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-:-KcUiSEB6231TWtWDPAD:b-eidnmu)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 4 of 7 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcUrEb8NXH_OU-v1c8j:-KcUrEb8NXH_OU-v1c8k:bn77tzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L246)):*
> ```Go
> 					// updates nl.Self() doesn't get clobbered by a heartbeat
> 					// that hasn't picked up the pause and gets the new value of
> 					// nl.Self().
> ```

I'm not sure I follow this comment, but it doesn't look like this eliminates a race, it just narrows it. I think there are too many independent synchronization mechanisms here - do we need to move all interaction with `pauseHeartbeat` under `heartbeatSem`? It's not clear to me why we have multiple semaphores in the first place - why do we want to allow concurrent increments and heartbeats?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 7 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/helpers_test.go, line 255 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcUs0eqIVV-pPTig_4F:-KcUs0eqIVV-pPTig_4G:b322y6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/helpers_test.go#L255)):*
> ```Go
> 
> // GetStores returns a copy of sl.stores.
> func (sl *StoreList) GetStores() []roachpb.StoreDescriptor {
> ```

Can this be `Stores`? We don't often use `GetFoo`.

---

*[pkg/storage/node_liveness.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOpv1PUNsaOG7EdtSi:-KcVLcDCUrD6SS0EnKze:b5lm7et) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L192)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Changed it to break out of the loop if the draining value is right. What exactly is not tested at the unit level?
</blockquote></details>

The behaviour of this function in the presence of a failed cput isn't tested, as far as I can tell.

---

*[pkg/storage/node_liveness.go, line 362 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcVMmAu7ycH2gv0z5IO:-KcVMmAu7ycH2gv0z5IP:b-n6cm0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L362)):*
> ```Go
> func (nl *NodeLiveness) setSelf(new Liveness) {
> 	nl.mu.Lock()
> 	defer nl.mu.Unlock()
> ```

nit: defers aren't free, and this one doesn't buy you much

---

*[pkg/storage/node_liveness.go, line 363 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcVMja6vMFWyp826UrH:-KcVMja6vMFWyp826UrI:b-9ix3wm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L363)):*
> ```Go
> 	nl.mu.Lock()
> 	defer nl.mu.Unlock()
> 	nl.mu.self = new
> ```

nit: new is a keyword, so probably better to use a different name

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/storage/node_liveness_test.go, line 462 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoJ3hfaKUNb0shQHq:-KcXXOrPpnhKgPD8aZ6e:bdqvrvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness_test.go#L462)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Modified to check presence/absence of the draining node's ID.
</blockquote></details>

Still, you're not directly testing the "target for any transfers".

---

*[pkg/storage/node_liveness_test.go, line 516 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcXX9Z131rGpc08ozzm:-KcXX9Z131rGpc08ozzn:b-jstqoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness_test.go#L516)):*
> ```Go
> 	// Stop and restart the store to verify that a restarted server clears the
> 	// draining field on the liveness record.
> 	mtc.stopStore(drainingNodeIdx)
> ```

can't you just set draining to false?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 4 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/storage/helpers_test.go, line 255 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcUs0eqIVV-pPTig_4F:-KcZ9RJx-Oi3A12z3HAL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/helpers_test.go#L255)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Can this be `Stores`? We don't often use `GetFoo`.
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOpv1PUNsaOG7EdtSi:-KcZxlT8cscPYeDh357e:b-g9ljr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The behaviour of this function in the presence of a failed cput isn't tested, as far as I can tell.
</blockquote></details>

Split this function up and tested this condition.

---

*[pkg/storage/node_liveness.go, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcUrEb8NXH_OU-v1c8j:-KcZF4beQB4rwwZRRvvR:b-x1tm27) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L246)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm not sure I follow this comment, but it doesn't look like this eliminates a race, it just narrows it. I think there are too many independent synchronization mechanisms here - do we need to move all interaction with `pauseHeartbeat` under `heartbeatSem`? It's not clear to me why we have multiple semaphores in the first place - why do we want to allow concurrent increments and heartbeats?
</blockquote></details>

I was thinking about the race in which a `SetDraining(true)` call pauses heartbeats, updates the record and a heartbeat (with `incrementEpoch=true`) then updates that record with `draining=false`. That can only happen because a heartbeat can observe the latest value of `nl.Self()` and proceed regardless of whether heartbeats are paused. I updated the comment, let me know if it still isn't clear.

Regarding a heartbeat happening at the same time as a `SetDraining(true)` call, I'm not sure anything unwanted would happen. A heartbeat setting `draining=false` will either succeed before `SetDraining(true)` in which case `SetDraining` will retry or fail if `SetDraining` cputs first because of its outdated copy of `nl.Self()`. I'm probably missing something though.

Regarding concurrent increments and heartbeats, I'm not sure. I talked to @spencerkimball and he mentioned that heartbeat code might have called increment epoch at some point so there was need for something reentrant. It could be good to simplify synchronization here since this isn't the case anymore.

---

*[pkg/storage/node_liveness.go, line 362 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcVMmAu7ycH2gv0z5IO:-KcZ9pCGUYzrK__CX8Cs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L362)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: defers aren't free, and this one doesn't buy you much
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness.go, line 363 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcVMja6vMFWyp826UrH:-KcZ9plvkynJSMRiQaf5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L363)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: new is a keyword, so probably better to use a different name
</blockquote></details>

Done.

---

*[pkg/storage/node_liveness_test.go, line 462 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcOoJ3hfaKUNb0shQHq:-KcZbtadZyBuvwUb2fqy:b-qj92pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/96aa17d31058efb89eee15c58fb7373ec2a1d517/pkg/storage/node_liveness_test.go#L462)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Still, you're not directly testing the "target for any transfers".
</blockquote></details>

Reworded.

---

*[pkg/storage/node_liveness_test.go, line 516 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcXX9Z131rGpc08ozzm:-KcZydyTOLvT0WpUg6ug:b-r51c9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness_test.go#L516)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can't you just set draining to false?
</blockquote></details>

I could just restart heartbeats but this is more in line with what all the other tests do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 5 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness_test.go, line 488 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-Kc_ZCJfMhpq7jrQ_EqQ:-Kc_ZCJfMhpq7jrQ_EqR:b-6n0nk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fdd871ecedcf98f7be08ef4341d04a69ea1f33/pkg/storage/node_liveness_test.go#L488)):*
> ```Go
> 	// given draining setting.
> 	if err := mtc.nodeLivenesses[drainingNodeIdx].SetDrainingInternal(ctx, &storage.Liveness{}, false); err != nil {
> 		t.Fatalf("unexpected error: %v", err)
> ```

this can just be `t.Fatal(err)` since err is non nil.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 5 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcUrEb8NXH_OU-v1c8j:-Kc_sSJdJeBnT63OMYcr:b-noinnz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L246)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I was thinking about the race in which a `SetDraining(true)` call pauses heartbeats, updates the record and a heartbeat (with `incrementEpoch=true`) then updates that record with `draining=false`. That can only happen because a heartbeat can observe the latest value of `nl.Self()` and proceed regardless of whether heartbeats are paused. I updated the comment, let me know if it still isn't clear.

Regarding a heartbeat happening at the same time as a `SetDraining(true)` call, I'm not sure anything unwanted would happen. A heartbeat setting `draining=false` will either succeed before `SetDraining(true)` in which case `SetDraining` will retry or fail if `SetDraining` cputs first because of its outdated copy of `nl.Self()`. I'm probably missing something though.

Regarding concurrent increments and heartbeats, I'm not sure. I talked to @spencerkimball and he mentioned that heartbeat code might have called increment epoch at some point so there was need for something reentrant. It could be good to simplify synchronization here since this isn't the case anymore.
</blockquote></details>

OK, so the check here is tied to the fact that SetDraining pauses heartbeats and updates Self in that order. That would work, although it's subtle for my tastes and I'd prefer to do both under one lock (reentrancy is rarely the best solution to a concurrency problem).

But there's a bigger problem here that I missed before: I don't think we should pause heartbeats while draining. If we pause heartbeats, then all our leases will abruptly expire at the end of one heartbeat interval (plus whatever grace period we use), which kind of defeats the purpose of draining. A node should be able to stay in draining mode and serve the leases it still holds until all the ranges have been drained away from it. It's very likely that that will take longer than one heartbeat interval. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-KcUrEb8NXH_OU-v1c8j:-KcsDI9gTvbaF-miv204:bgcqka8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cdcc6bae1df2686df98681a58a74ae6a1b8cc85/pkg/storage/node_liveness.go#L246)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

OK, so the check here is tied to the fact that SetDraining pauses heartbeats and updates Self in that order. That would work, although it's subtle for my tastes and I'd prefer to do both under one lock (reentrancy is rarely the best solution to a concurrency problem).

But there's a bigger problem here that I missed before: I don't think we should pause heartbeats while draining. If we pause heartbeats, then all our leases will abruptly expire at the end of one heartbeat interval (plus whatever grace period we use), which kind of defeats the purpose of draining. A node should be able to stay in draining mode and serve the leases it still holds until all the ranges have been drained away from it. It's very likely that that will take longer than one heartbeat interval. 
</blockquote></details>

Good point. I moved everything under one lock and removed the logic to pause heartbeats.

---

*[pkg/storage/node_liveness_test.go, line 488 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-Kc_ZCJfMhpq7jrQ_EqQ:-Kcs3etom0n_LalZPK_U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fdd871ecedcf98f7be08ef4341d04a69ea1f33/pkg/storage/node_liveness_test.go#L488)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this can just be `t.Fatal(err)` since err is non nil.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 7 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11724#-:-Kd1_Xe4mx0gI1lUHyJM:bnfp4nl)*
<!-- Sent from Reviewable.io -->
